[hop free symbols][refactor] make create_graph_input always take example_value

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #138363
* #139663
* #138559
* #138737
* #138558
* __->__ #138428
* #138345

Code refactoring only. We move the wrap_to_fake_tensor_logic out of wrap_fx_proxy for placeholders to provide the invariant that **all graph inputs must set their example values when creating the inputs**. This invariant helps us to identify all the free symbols in the graph in top-level and sub-graphs.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

ydwu4 committed:
[hop free symbols][refactor] make create_graph_input always take example_value

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop free symbols][refactor] make create_graph_input always take example_value"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop free symbols][refactor] make create_graph_input always take example_value"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop free symbols][refactor] make create_graph_input always take example_value"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

ydwu4 commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1707,15 +1711,22 @@ def wrap_numpy_ndarray(self, value):
+        example_value = wrap_to_fake_tensor_and_record(
+            tensor_value,
+            tx=self.tx,
+            is_tensor=False,
```
is_tensor is dermined by whether target_cls in (TensorVariable, TensorWithTFOverrideVariable). NumpyNdarraryVaraible is not in them so is_tensor = False. This PR keeps this behavior unchanged and only do code-refactoring.

ydwu4 commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1927,17 +1937,21 @@ def wrap_symfloat(self, value):
         # TODO: Maybe the tensor-ification should be built into the source,
         # rather than by special pattern match
+        example_value = wrap_to_fake_tensor_and_record(
+            wrapped_value, tx=self.tx, is_tensor=False, source=self.get_source()
```
same here for is_tensor=False

ydwu4 commented on torch/_dynamo/variables/builder.py:
```diff
@@ -2001,17 +2015,21 @@ def wrap_unspecialized_primitive(self, value):
         options.update({"raw_value": value})
 
+        example_value = wrap_to_fake_tensor_and_record(
+            wrapped_value, tx=self.tx, is_tensor=False, source=self.get_source()
```
same here

ydwu4 commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1618,10 +1605,27 @@ def wrap_tensor(self, value: torch.Tensor):
+        # the subgraph.
+        # See NOTE [HigherOrderOperator tracing design] for more details.
+
+        example_value = wrap_to_fake_tensor_and_record(
```
We have to move the wrap_to_fake_tensor_and_record logic in wrap_fx_proxy out, so that we could provide an example_value whose symints can later be tracked.

ydwu4 committed:
Update on "[hop free symbols][refactor] make create_graph_input always take example_value"


Code refactoring only. We move the wrap_to_fake_tensor_logic out of wrap_fx_proxy for placeholders to provide the invariant that **all graph inputs must have example values**


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop free symbols][refactor] make create_graph_input always take example_value"


Code refactoring only. We move the wrap_to_fake_tensor_logic out of wrap_fx_proxy for placeholders to provide the invariant that **all graph inputs must have example values**


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

ydwu4 commented on test/dynamo/test_autograd_function.py:
```diff
@@ -546,14 +546,14 @@ def forward(self, L_x_: "f32[]", L_z_: "f32[]", L_weird_b: "f32[]", L_weird_c: "
 
     class fwd_body_0(torch.nn.Module):
-        def forward(self, ctx, x: "f32[]", z: "f32[]", l_weird_b: "f32[]", l_weird_c: "f32[]"):
+        def forward(self, ctx : torch.autograd.function.Function, x: "f32[]", z: "f32[]", l_weird_b: "f32[]", l_weird_c: "f32[]"):
```
The context is a torch.autograd.funciton.Function instead of https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/side_effects.py#L270

zou3519 replied:
What causes some of these to state their types but some of them not to?

ydwu4 committed:
Update on "[hop free symbols][refactor] make create_graph_input always take example_value"


Code refactoring only. We move the wrap_to_fake_tensor_logic out of wrap_fx_proxy for placeholders to provide the invariant that **all graph inputs must set their example values when creating the inputs**. This invariant helps us to identify all the free symbols in the graph in top-level and sub-graphs.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop free symbols][refactor] make create_graph_input always take example_value"


Code refactoring only. We move the wrap_to_fake_tensor_logic out of wrap_fx_proxy for placeholders to provide the invariant that **all graph inputs must set their example values when creating the inputs**. This invariant helps us to identify all the free symbols in the graph in top-level and sub-graphs.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop free symbols][refactor] make create_graph_input always take example_value"


Code refactoring only. We move the wrap_to_fake_tensor_logic out of wrap_fx_proxy for placeholders to provide the invariant that **all graph inputs must set their example values when creating the inputs**. This invariant helps us to identify all the free symbols in the graph in top-level and sub-graphs.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

zou3519 commented on test/functorch/test_control_flow.py:
```diff
@@ -3150,15 +3150,15 @@ def forward(self, L_iter_ : torch.Tensor, L_x_ : torch.Tensor, L_self_buffers_de
                 gm.cond_fn_0.code.strip(),
                 """\
-def forward(self, l_iter_, l_x_, l_self_buffers_dec__cond_fn, l_self_modules_linear_parameters_bias__body_fn, l_self_modules_linear_parameters_weight__body_fn):
+def forward(self, l_iter_ : torch.Tensor, l_x_ : torch.Tensor, l_self_buffers_dec__cond_fn, l_self_modules_linear_parameters_bias__body_fn, l_self_modules_linear_parameters_weight__body_fn):
```
Why do the `l_self_buffers_dec__cond_fn` don't have types?

zou3519 commented on torch/_dynamo/output_graph.py:
```diff
@@ -443,11 +443,14 @@ def get_backward_state_proxy(self):
         if self.backward_state_proxy is None:
             if self.export:
                 unimplemented("backward_state does not support export")
+            example_value = BackwardState()
```
Can you check (add a test?) that this doesn't cause a warning? Direct initialization of FunctionCtx usually induces a warniing

zou3519 commented on torch/_dynamo/output_graph.py:
```diff
@@ -670,12 +673,12 @@ def bind_symint(s: torch.SymInt, prop):
             proxy = self.root_tracer.create_graph_input(
                 str(s0),
-                torch.SymInt,
+                type(s),
```
Is there a reason to pass type(s) if there's always an example value? create_graph_input can always get the type from the example value.

zou3519 reviewed (approved):
Can you test that creating the backward state does not add user-facing warnings? We've had that problem in the past.

Other than that, this looks fine to me

ydwu4 committed:
Update on "[hop free symbols][refactor] make create_graph_input always take example_value"


Code refactoring only. We move the wrap_to_fake_tensor_logic out of wrap_fx_proxy for placeholders to provide the invariant that **all graph inputs must set their example values when creating the inputs**. This invariant helps us to identify all the free symbols in the graph in top-level and sub-graphs.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop free symbols][refactor] make create_graph_input always take example_value"


Code refactoring only. We move the wrap_to_fake_tensor_logic out of wrap_fx_proxy for placeholders to provide the invariant that **all graph inputs must set their example values when creating the inputs**. This invariant helps us to identify all the free symbols in the graph in top-level and sub-graphs.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]
